django/core/mail/utils.py

```diff
@@ -5,6 +5,18 @@
 import socket
 
 
+def punycode(domain):
+    """
+    Return the Punycode of the given domain if it's non-ASCII. The given
+    domain, otherwise.
+    """
+    try:
+        domain.encode('ascii')
+    except UnicodeEncodeError:
+        return domain.encode('idna').decode('ascii')
```

Comment:
Looking at other instances of the string `idna` it seems like it can raise some form of `UnicodeError`.

https://github.com/django/django/blob/fbb83fefd43a89d1a802d77dcc3a9c9ac15a5afa/django/utils/html.py#L320-L324

https://github.com/django/django/blob/42b9a23267f14be39b9b00958e18d5746783208e/django/core/validators.py#L126-L129

These instances should probably use this function as well and I think `django.utils.encoding` could be a great home for it.

Reply:
Agreed, we can move it to `django.utils.encoding`, but I don't think that we should catch `UnicodeError` in this hook.